Description: Fix segfaults
Author: Michael Meskes <meskes@debian.org>

--- bsdmainutils/usr.bin/col/col.c	2017-12-19 16:24:54.197178946 +0100
+++ bsdmainutils/usr.bin/col/col.c	2017-12-19 16:28:08.117289280 +0100
@@ -82,7 +80,7 @@
 typedef struct char_str {
 #define	CS_NORMAL	1
 #define	CS_ALTERNATE	2
-	short		c_column;	/* column character is in */
+	int		c_column;	/* column character is in */
 	CSET		c_set;		/* character set (currently only 2) */
 	wchar_t		c_char;		/* character in question */
 	int		c_width;	/* character width */
@@ -229,7 +221,8 @@
 				cur_col = 0;
 				continue;
 			case SPACE:
-				++cur_col;
+				if (cur_col < INT_MAX)
+					++cur_col;
 				continue;
 			case SI:
 				cur_set = CS_NORMAL;
@@ -238,8 +231,10 @@
 				cur_set = CS_ALTERNATE;
 				continue;
 			case TAB:		/* adjust column */
-				cur_col |= 7;
-				++cur_col;
+				if (cur_col < INT_MAX) {
+					cur_col |= 7;
+					++cur_col;
+				}
 				continue;
 			case VT:
 				addto_lineno(&cur_line, -2);
@@ -335,7 +330,7 @@
 			l->l_needs_sort = 1;
 		else
 			l->l_max_col = cur_col;
-		if (c->c_width > 0)
+		if (c->c_width > 0 && cur_col < INT_MAX)
 			cur_col += c->c_width;
 	}
 	if (ferror(stdin))
