Description: Remove incompatible getopt usage.
Author: Michael Meskes <meskes@debian.org>

--- freebsd/usr-bin/colcrt/colcrt.c	2009-09-23 14:51:28.000000000 +0200
+++ bsdmainutils/usr.bin/colcrt/colcrt.c	2009-09-23 14:51:34.000000000 +0200
@@ -90,19 +89,22 @@
 
 	setlocale(LC_ALL, "");
 
-	while ((ch = getopt(argc, argv, "-2")) != -1)
-		switch (ch) {
-		case '-':
-			suppresul = 1;
-			break;
-		case '2':
-			printall = 1;
-			break;
-		default:
-			usage();
-		}
-	argc -= optind;
-	argv += optind;
+	/* do not use getopt() here because the GNU version differs
+	 * significantly from the BSD version when using an option name "-" */
+	for (argv++, argc--; *argv; argv++, argc--) {
+                if (!strcmp (*argv, "-"))
+                        suppresul = 1;
+                else if (!strcmp (*argv, "-2"))
+                        printall = 1;
+                else if (!strcmp (*argv, "--")) {
+                        argv++;
+                        argc--;
+                        break;
+                } else if (**argv == '-')
+                        usage();
+                else
+                        break;
+        }
 
 	do {
 		if (argc > 0) {
