Description: Make program compile on Debian.
Author: Michael Meskes <meskes@debian.org>

--- freebsd/usr.bin/hexdump/conv.c	2009-10-03 18:45:29.000000000 +0200
+++ bsdmainutils/usr.bin/hexdump/conv.c	2009-10-03 18:46:04.000000000 +0200
@@ -35,7 +35,6 @@
 static const char sccsid[] = "@(#)conv.c	8.1 (Berkeley) 6/6/93";
 #endif /* not lint */
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD$");
 
 #include <sys/types.h>
 
--- freebsd/usr.bin/hexdump/display.c	2009-10-03 18:45:29.000000000 +0200
+++ bsdmainutils/usr.bin/hexdump/display.c	2009-10-03 18:46:12.000000000 +0200
@@ -37,7 +37,6 @@
 #endif
 #endif /* not lint */
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD$");
 
 #include <sys/param.h>
 #include <sys/stat.h>
--- freebsd/usr.bin/hexdump/hexdump.c	2009-10-03 18:45:29.000000000 +0200
+++ bsdmainutils/usr.bin/hexdump/hexdump.c	2009-10-03 18:46:23.000000000 +0200
@@ -43,7 +43,6 @@
 #endif
 #endif /* not lint */
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD$");
 
 #include <sys/types.h>
 #include <locale.h>
--- freebsd/usr.bin/hexdump/odsyntax.c	2009-10-03 18:45:29.000000000 +0200
+++ bsdmainutils/usr.bin/hexdump/odsyntax.c	2009-10-03 18:46:36.000000000 +0200
@@ -37,7 +37,6 @@
 #endif
 #endif /* not lint */
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD$");
 
 #include <sys/types.h>
 
@@ -311,7 +310,7 @@
 }
 
 static const char *
-odformatfp(char fchar __unused, const char *fmt)
+odformatfp(char fchar /* unused */, const char *fmt)
 {
 	size_t isize;
 	int digits;
--- freebsd/usr.bin/hexdump/parse.c	2009-10-03 18:45:29.000000000 +0200
+++ bsdmainutils/usr.bin/hexdump/parse.c	2009-10-03 18:46:44.000000000 +0200
@@ -37,7 +37,6 @@
 #endif
 #endif /* not lint */
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD$");
 
 #include <sys/types.h>
 
@@ -142,7 +141,8 @@
 				badfmt(fmt);
 		if (!(tfu->fmt = malloc(p - savep + 1)))
 			err(1, NULL);
-		(void) strlcpy(tfu->fmt, savep, p - savep + 1);
+		(void) strncpy(tfu->fmt, savep, p - savep + 1);
+		(tfu->fmt)[p - savep] = '\0'; /* make sure it's null terminated */
 		escape(tfu->fmt);
 		p++;
 	}
--- freebsd/usr.bin/hexdump/hexsyntax.c	2009-10-03 18:45:29.000000000 +0200
+++ bsdmainutils/usr.bin/hexdump/hexsyntax.c	2009-10-03 18:50:47.000000000 +0200
@@ -37,7 +37,6 @@
 #endif
 #endif /* not lint */
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD$");
 
 #include <sys/types.h>
 
