Description: Add binary search option.
Author: Michael Meskes <meskes@debian.org>

--- bsdmainutils/usr.bin/look/look.1	2017-12-23 12:36:18.920572190 +0100
+++ bsdmainutils/usr.bin/look/look.1	2017-12-23 12:41:33.282875551 +0100
@@ -36,7 +36,7 @@
 .Nd display lines beginning with a given string
 .Sh SYNOPSIS
 .Nm
-.Op Fl df
+.Op Fl bdf
 .Op Fl t Ar termchar
 .Ar string
 .Op Ar
@@ -48,11 +48,6 @@
 which contain
 .Ar string
 as a prefix.
-As
-.Nm
-performs a binary search, the lines in
-.Ar file
-must be sorted.
 .Pp
 If
 .Ar file
@@ -63,6 +58,15 @@
 .Pp
 The following options are available:
 .Bl -tag -width indent
+.It Fl b, -binary
+Use a binary search on the given word list. If you are ignoring case with
+.Fl f
+or ignoring non-alphanumeric characters with
+.Fl d ,
+the file must be sorted in the same way. Please note that these options are the
+default if no filename is given. See
+.Xr sort 1
+for more information on sorting files. 
 .It Fl d , -alphanum
 Dictionary character set and order, i.e., only alphanumeric characters
 are compared.
@@ -104,6 +108,10 @@
 This was incorrect and the current man page matches the historic
 implementation.
 .Pp
+.Nm
+uses a linear search by default instead of a binary search, which is what most
+other implementations use by default.
+.Pp
 The
 .Fl a
 and
--- bsdmainutils/usr.bin/look/look.c	2017-12-23 12:36:27.000000000 +0100
+++ bsdmainutils/usr.bin/look/look.c	2017-12-23 12:40:52.153710471 +0100
@@ -80,7 +80,7 @@
 #define	GREATER		1
 #define	LESS		(-1)
 
-static int dflag, fflag;
+static int dflag, fflag, bflag;
 
 static char	*binary_search(wchar_t *, unsigned char *, unsigned char *);
 static int	 compare(wchar_t *, unsigned char *, unsigned char *);
@@ -88,11 +88,13 @@
 static int	 look(wchar_t *, unsigned char *, unsigned char *);
 static wchar_t	*prepkey(const char *, wchar_t);
 static void	 print_from(wchar_t *, unsigned char *, unsigned char *);
+static int      grep(wchar_t *, unsigned char *, unsigned char *);
 
 static void usage(void);
 
 static struct option longopts[] = {
 	{ "alternative",no_argument,	NULL, 'a' },
+	{ "binary",	no_argument,	NULL, 'b' },
 	{ "alphanum",	no_argument,	NULL, 'd' },
 	{ "ignore-case",no_argument,	NULL, 'i' },
 	{ "terminate",	required_argument, NULL, 't'},
@@ -113,11 +115,14 @@
 
 	file = _path_words;
 	termchar = L'\0';
-	while ((ch = getopt_long(argc, argv, "+adft:", longopts, NULL)) != -1)
+	while ((ch = getopt_long(argc, argv, "+abdft:", longopts, NULL)) != -1)
 		switch(ch) {
 		case 'a':
 			/* COMPATIBILITY */
 			break;
+		case 'b':
+			bflag = 1;
+			break;
 		case 'd':
 			dflag = 1;
 			break;
@@ -158,7 +163,10 @@
 		if ((front = mmap(NULL, (size_t)sb.st_size, PROT_READ, MAP_SHARED, fd, (off_t)0)) == MAP_FAILED)
 			err(2, "%s", file);
 		back = front + sb.st_size;
-		match *= (look(key, front, back));
+		if (bflag)
+                        match *= (look(key, front, back));
+                else
+                        match *= (grep(key, front, back));
 		close(fd);
 	} while (argc-- > 2 && (file = *argv++));
 
@@ -302,6 +310,32 @@
 }
 
 /*
+ * Look at every string in the wordlist for a match. Since wordlists are often
+ * not reliably sorted, or are sorted in locale-dependent ways, we use this
+ * method by default.
+ */
+int
+grep(wchar_t *string, unsigned char *front, unsigned char *back)
+{
+	int ret = 1;
+
+	for (; front < back; ++front) {
+		if (compare(string, front, back) == EQUAL) {
+			for (; front < back && *front != '\n'; ++front)
+				if (putchar(*front) == EOF)
+					err(2, "stdout");
+			if (putchar('\n') == EOF)
+				err(2, "stdout");
+			ret = 0;
+		}
+
+		/* Move to the next word in the list. */
+                while (front < back && *front != '\n')
+                        front++;
+	}
+	return(ret);
+}
+/*
  * Print as many lines as match string, starting at front.
  */
 static void
@@ -358,6 +392,6 @@
 static void
 usage(void)
 {
-	(void)fprintf(stderr, "usage: look [-df] [-t char] string [file ...]\n");
+	(void)fprintf(stderr, "usage: look [-bdf] [-t char] string [file ...]\n");
 	exit(2);
 }
