Description: Use signal names for POSIX compliance.
Author: Michael Meskes <meskes@debian.org>

--- bsdmainutils/usr.bin/lorder/lorder.sh.orig	2016-09-29 12:05:45.948371887 +0200
+++ bsdmainutils/usr.bin/lorder/lorder.sh	2016-09-29 12:06:27.012392213 +0200
@@ -48,7 +48,7 @@
 NM=${NM:-nm}
 
 # remove temporary files on HUP, INT, QUIT, PIPE, TERM
-trap "rm -f $R $S $T; exit 1" 1 2 3 13 15
+trap "rm -f $R $S $T; exit 1" HUP INT QUIT PIPE TERM
 
 # make sure all the files get into the output
 for i in $*; do
