Description: Fix sort order.
Author: Michael Meskes <meskes@debian.org>

--- bsd/openbsd/lorder/lorder.sh	2009-10-17 17:06:44.000000000 +0200
+++ bsdmainutils/usr.bin/lorder/lorder.sh	2009-10-17 17:22:15.000000000 +0200
@@ -75,7 +75,7 @@
 
 # sort symbols and references on the first field (the symbol)
 # join on that field, and print out the file names (dependencies).
-sort +1 $R -o $R
-sort +1 $S -o $S
+sort -k 2 $R -o $R
+sort -k 2 $S -o $S
 join -j 2 -o 1.1 2.1 $R $S
 rm -f $R $S
