Description: Add option to set first day of week.
Author: Michael Meskes <meskes@debian.org>

--- bsdmainutils/usr.bin/ncal/ncal.c	2011-01-03 13:50:39.000000000 +0100
+++ bsdmainutils/usr.bin/ncal/ncal.c	2011-01-03 13:52:27.000000000 +0100
@@ -273,7 +273,7 @@
 #endif
 	before = after = -1;
 
-	while ((ch = getopt(argc, argv, "3A:B:Cd:eH:hjJm:Nops:wy")) != -1)
+	while ((ch = getopt(argc, argv, "3A:B:Cd:eH:hjJm:Nops:wySM")) != -1)
 		switch (ch) {
 		case '3':
 			flag_3months = 1;
@@ -362,6 +362,16 @@
 		case 'y':
 			flag_wholeyear = 1;
 			break;
+		case 'S':
+			if (flag_backward)
+				usage();
+			weekstart = 0;
+			break;
+		case 'M':
+			if (flag_backward)
+				usage();
+			weekstart = 1;
+			break;
 		default:
 			usage();
 		}
@@ -528,8 +538,8 @@
 	fputs(
 "Usage: cal [general options] [-hjy] [[month] year]\n"
 "       cal [general options] [-hj] [-m month] [year]\n"
-"       ncal [general options] [-hJjpwy] [-s country_code] [[month] year]\n"
-"       ncal [general options] [-hJeo] [year]\n"
+"       ncal [general options] [-hJjpwySM] [-s country_code] [[month] year]\n"
+"       ncal [general options] [-hJeoSM] [year]\n"
 "General options: [-NC3] [-A months] [-B months]\n"
 "For debug the highlighting: [-H yyyy-mm-dd] [-d yyyy-mm]\n",
 	    stderr);
--- bsdmainutils/usr.bin/ncal/ncal.1	2011-01-03 13:52:56.000000000 +0100
+++ bsdmainutils/usr.bin/ncal/ncal.1	2011-01-03 13:54:07.000000000 +0100
@@ -47,7 +47,7 @@
 .Fl m Ar month
 .Op Ar year
 .Nm ncal
-.Op Fl 3hjJpwy
+.Op Fl 3hjJpwySM
 .Op Fl A Ar number
 .Op Fl B Ar number
 .Op Fl s Ar country_code
@@ -56,7 +56,7 @@
 .Ar year
 .Oc
 .Nm ncal
-.Op Fl 3hJeo
+.Op Fl 3hJeoSM
 .Op Fl A Ar number
 .Op Fl B Ar number
 .Op Ar year
@@ -147,6 +147,10 @@
 Use
 .Ar yyyy-mm-dd
 as the current date (for debugging of highlighting).
+.It Fl M
+Weeks start on Monday.
+.It Fl S
+Weeks start on Sunday.
 .El
 .Pp
 A single parameter specifies the year (1\(en9999) to be displayed;
