Description: Remove capsicum code
Author: Michael Meskes <meskes@debian.org>

--- bsdmainutils/usr.bin/write/write.c	2017-12-20 11:37:51.796262348 +0100
+++ bsdmainutils/usr.bin/write/write.c	2017-12-20 12:54:59.348830683 +0100
@@ -48,13 +48,10 @@
 __FBSDID("$FreeBSD: head/usr.bin/write/write.c 326025 2017-11-20 19:49:47Z pfg $");
 
 #include <sys/param.h>
-#include <sys/capsicum.h>
-#include <sys/filio.h>
 #include <sys/signal.h>
 #include <sys/stat.h>
 #include <sys/time.h>
 
-#include <capsicum_helpers.h>
 #include <ctype.h>
 #include <err.h>
 #include <errno.h>
@@ -80,8 +80,6 @@
 int
 main(int argc, char **argv)
 {
-	unsigned long cmds[] = { TIOCGETA, TIOCGWINSZ, FIODGNAME };
-	cap_rights_t rights;
 	struct passwd *pwd;
 	time_t atime;
 	uid_t myuid;
@@ -95,31 +93,6 @@
 	devfd = open(_PATH_DEV, O_RDONLY);
 	if (devfd < 0)
 		err(1, "open(/dev)");
-	cap_rights_init(&rights, CAP_FCNTL, CAP_FSTAT, CAP_IOCTL, CAP_LOOKUP,
-	    CAP_PWRITE);
-	if (cap_rights_limit(devfd, &rights) < 0 && errno != ENOSYS)
-		err(1, "can't limit devfd rights");
-
-	/*
-	 * Can't use capsicum helpers here because we need the additional
-	 * FIODGNAME ioctl.
-	 */
-	cap_rights_init(&rights, CAP_FCNTL, CAP_FSTAT, CAP_IOCTL, CAP_READ,
-	    CAP_WRITE);
-	if ((cap_rights_limit(STDIN_FILENO, &rights) < 0 && errno != ENOSYS) ||
-	    (cap_rights_limit(STDOUT_FILENO, &rights) < 0 && errno != ENOSYS) ||
-	    (cap_rights_limit(STDERR_FILENO, &rights) < 0 && errno != ENOSYS) ||
-	    (cap_ioctls_limit(STDIN_FILENO, cmds, nitems(cmds)) < 0 && errno != ENOSYS) ||
-	    (cap_ioctls_limit(STDOUT_FILENO, cmds, nitems(cmds)) < 0 && errno != ENOSYS) ||
-	    (cap_ioctls_limit(STDERR_FILENO, cmds, nitems(cmds)) < 0 && errno != ENOSYS) ||
-	    (cap_fcntls_limit(STDIN_FILENO, CAP_FCNTL_GETFL) < 0 && errno != ENOSYS) ||
-	    (cap_fcntls_limit(STDOUT_FILENO, CAP_FCNTL_GETFL) < 0 && errno != ENOSYS) ||
-	    (cap_fcntls_limit(STDERR_FILENO, CAP_FCNTL_GETFL) < 0 && errno != ENOSYS))
-		err(1, "can't limit stdio rights");
-
-	caph_cache_catpages();
-	caph_cache_tzdata();
-
 	/*
 	 * Cache UTX database fds.
 	 */
@@ -137,9 +110,6 @@
 			login = "???";
 	}
 
-	if (cap_enter() < 0 && errno != ENOSYS)
-		err(1, "cap_enter");
-
 	while (getopt(argc, argv, "") != -1)
 		usage();
 	argc -= optind;
